internal/pkg/docker/orchestrator/orchestrator.go

```diff
@@ -452,6 +486,88 @@ func (o *Orchestrator) waitForContainerToStart(ctx context.Context, id string) e
 	}
 }
 
+func (o *Orchestrator) waitForContainerDependencies(ctx context.Context, name string, definitions map[string]ContainerDefinition) error {
+	var deps []string
+	for depName, state := range definitions[name].DependsOn {
+		deps = append(deps, fmt.Sprintf("%s->%s", depName, state))
+	}
+	logMsg := strings.Join(deps, ", ")
+	fmt.Printf("Waiting for container %q dependencies: [%s]\n", name, color.Emphasize(logMsg))
+	eg, ctx := errgroup.WithContext(ctx)
+	for name, state := range definitions[name].DependsOn {
+		name, state := name, state
+		eg.Go(func() error {
+			ctrId := o.containerID(name)
+			isEssential := definitions[name].IsEssential
```

Comment:
To me, it's a bit counter-intuitive that a container would care about whether the containers that it depends on is essential 💭 For example, if the `depends_on` specify that container A can start only after container B has started, why would container A care whether container B is essential?

It seems like we are using `isEssential` here majorly to decide:
1. whether to error out from `waitForContainerToStart` 
2. whether to `log.Successf` or to just `fmt.Printf`
3. whether to error out from `IsContainerHealthy`


For 1, would https://github.com/aws/copilot-cli/blob/08542d3c0bcf62dc0b894e732039c41fccf92051/internal/pkg/docker/orchestrator/orchestrator.go#L246 would have been a better place to handle whether to error out or not depending on the essential-ity? Then, we will swallow the "ErrContainerExited` error here (regardless of whether it is essential), because if the contaienr has exited, that means it must have been started.

For 2, I think  https://github.com/aws/copilot-cli/blob/08542d3c0bcf62dc0b894e732039c41fccf92051/internal/pkg/docker/orchestrator/orchestrator.go#L246 again is the better place to have that logic.

For 3, I have the same question in https://github.com/aws/copilot-cli/pull/5509/files#r1428611649.

Essentially, I feel like we are mixing two routes of logics here:
1. To complain or just log when an essential/non-essential contaienr is not functioning
2. To determine whether to start a container based on its dependency
The function that handles 2, I think, typically should not care about the essential-ity of the containers.


Let me know what you think!

Reply:
Thank you for pointing me about this. yeah, essential-ity and dependsOn are different entities.
I changed the logic to reflect the above.